### ISSUE TITLE
[notifications][android] addPushTokenListener should called after request token

### ISSUE THREAD

**Issue by dulyts** [expo/expo#29909]
### Minimal reproducible example

https://github.com/dulyts/expo-notifications-issue

### What platform(s) does this occur on?

Android

### Did you reproduce this issue in a development build?

Yes

### Summary

In previous SDKs Notifications.addPushTokenListener callback was called after request push token with Notifications.getDevicePushTokenAsync

### Environment

  expo-env-info 1.2.0 environment info:
    System:
      OS: Windows 10 10.0.19045
    Binaries:
      Node: 20.11.1 - C:\Program Files\nodejs\node.EXE
      Yarn: 1.19.1 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
      npm: 10.2.4 - C:\Program Files\nodejs\npm.CMD
      Watchman: 20220918.223204.0 - C:\ProgramData\chocolatey\bin\watchman.EXE
    SDKs:
      Android SDK:
        API Levels: 28, 29, 30, 31, 33, 34
        Build Tools: 34.0.0
        System Images: android-31 | Google APIs Intel x86_64 Atom
    IDEs:
      Android Studio: AI-212.5712.43.2112.8815526
    npmPackages:
      expo: ~51.0.14 => 51.0.14
      react: 18.2.0 => 18.2.0
      react-native: 0.74.2 => 0.74.2
    Expo Workflow: managed

### Expo Doctor Diagnostics

✔ Check Expo config for common issues
✔ Check package.json for common issues
✔ Check native tooling versions
✔ Check dependencies for packages that should not be installed directly
✔ Check for common project setup issues
✔ Check npm/ yarn versions
✔ Check for issues with metro config
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check that native modules do not use incompatible support packages
✔ Check for legacy global CLI installed locally
✔ Check that native modules use compatible support package versions for installed Expo SDK
✔ Check that packages match versions required by installed Expo SDK

Didn't find any issues with the project!

**Comment by GrinZero** [comment]
Can you explain the necessity of this trigger? Is it for compatibility? Because I am using the new version directly, I believe the current logic is acceptable

**Comment by dulyts** [comment]
Based on the code it should called (https://github.com/expo/expo/blob/sdk-51/packages/expo-notifications/android/src/main/java/expo/modules/notifications/tokens/PushTokenModule.kt#L61) but for something it did not happening.
And also it works different on iOS and android devices, because when I call getDevicePushTokenAsync listener callback is called.
For the token registration I only use the listener (because if OS replace the token I can handle the new one) to call backend for save token. When user enable notification I request a push token with getDevicePushTokenAsync, and the listener do everthing else.

**Comment by GrinZero** [comment]
> Based on the code it should called (https://github.com/expo/expo/blob/sdk-51/packages/expo-notifications/android/src/main/java/expo/modules/notifications/tokens/PushTokenModule.kt#L61) but for something it did not happening. And also it works different on iOS and android devices, because when I call getDevicePushTokenAsync listener callback is called. For the token registration I only use the listener (because if OS replace the token I can handle the new one) to call backend for save token. When user enable notification I request a push token with getDevicePushTokenAsync, and the listener do everthing else.

Okay, it can be confirmed that there is indeed an issue of inconsistent performance between the two platforms, which should be a defect. I will try to fix it.

**Comment by dulyts** [comment]
@GrinZero are you tried to fix it?

**Comment by GrinZero** [comment]
I'm very sorry, after our discussion that day, I briefly reviewed the code and did not provide a fix. I have been working on my other projects recently and don't have time to fix this issue.

**Comment by scarlac** [comment]
This is still a known issues with Expo v52.0.31.

@GrinZero
The inconsistency between the platforms is causing regression for our app.
Specifically, if you simply want the token, but also want to know when it changes, while the app is running, you end up with 2 events: 1 for the initial read + 1 for the listener that gets triggered on iOS.
But on Android, you only get the initial read. You are now faced with potential bugs or you have to design your system so it's idempotent, or so that it ignores the duplicate events depending on the platform.

**Comment by svipas** [comment]
@scarlac I totally agree with you, it would be amazing, because now simple thing became a bit of mess, because of those additional checks and gotchas we have to do
